azure/scope/clients.go

```diff
@@ -169,3 +205,151 @@ func setValue(settings auth.EnvironmentSettings, key string) {
+			Kind:       kind,
+			Name:       asoClientSecret.AzureClusterIdentity.GetName(),
+			UID:        asoClientSecret.AzureClusterIdentity.GetUID(),
+		}
+
+		// Fetch identity secret, if it exists
+		key := types.NamespacedName{
+			Namespace: asoClientSecret.AzureClusterIdentity.Spec.ClientSecret.Namespace,
+			Name:      asoClientSecret.AzureClusterIdentity.Spec.ClientSecret.Name,
+		}
+		identitySecret := &corev1.Secret{}
+		err := asoClientSecret.KubeClient.Get(ctx, key, identitySecret)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to fetch identity secret %s as defined in AzureClusterIdentity %q/%q: %w", identitySecret.Name, asoClientSecret.AzureClusterIdentity.Namespace, asoClientSecret.AzureClusterIdentity.Name, err)
+		}
+
+		switch asoClientSecret.AzureClusterIdentity.Spec.Type {
+		case infrav1.ServicePrincipal, infrav1.ManualServicePrincipal:
+			newASOSecret.Data["AZURE_CLIENT_SECRET"] = identitySecret.Data["clientSecret"]
+		case infrav1.ServicePrincipalCertificate:
+			newASOSecret.Data["AZURE_CLIENT_CERTIFICATE"] = identitySecret.Data["certificate"]
+			newASOSecret.Data["AZURE_CLIENT_CERTIFICATE_PASSWORD"] = identitySecret.Data["password"]
+		case infrav1.UserAssignedMSI:
+			newASOSecret.Data["AZURE_IDENTITY_RESOURCE_ID"] = []byte(asoClientSecret.AzureClusterIdentity.Spec.ResourceID)
+		}
+	} else {
+		newASOSecret.ObjectMeta = metav1.ObjectMeta{
+			Name:      aso.GetASOSecretName(asoClientSecret.ClusterObject.GetName(), asoClientSecret.SubscriptionID),
+			Namespace: asoClientSecret.ClusterObject.GetNamespace(),
+		}
+
+		newASOSecret.Data["AZURE_TENANT_ID"] = []byte(c.TenantID())
+		newASOSecret.Data["AZURE_CLIENT_ID"] = []byte(c.ClientID())
+
+		gvk := asoClientSecret.ClusterObject.GetObjectKind().GroupVersionKind()
+		owner = metav1.OwnerReference{
+			APIVersion: gvk.GroupVersion().String(),
+			Kind:       gvk.Kind,
+			Name:       asoClientSecret.ClusterObject.GetName(),
+			UID:        asoClientSecret.ClusterObject.GetUID(),
+		}
+
+		// 1.Client Credentials
+		if _, e := c.GetClientCredentials(); e == nil {
+			newASOSecret.Data["AZURE_CLIENT_SECRET"] = []byte(c.ClientSecret())
+		}
+
+		// 2. Client Certificate
+		if _, e := c.GetClientCertificate(); e == nil {
+			cert, err := getCertificateFromFile(c.Values[auth.CertificatePath])
+			if err != nil {
+				return err
+			}
+			newASOSecret.Data["AZURE_CLIENT_CERTIFICATE"] = cert
+			newASOSecret.Data["AZURE_CLIENT_CERTIFICATE_PASSWORD"] = []byte(c.Values[auth.CertificatePassword])
+		}
+
+		// 3. MSI
+		if !adal.MSIAvailable(context.Background(), nil) {
+			return fmt.Errorf("MSI not available")
+		} else {
+			newASOSecret.Data["AZURE_IDENTITY_RESOURCE_ID"] = []byte(c.Values[auth.Resource])
+		}
+	}
+
+	return reconcileSecret(ctx, asoClientSecret.KubeClient, newASOSecret, owner)
+}
+
+func reconcileSecret(ctx context.Context, kubeClient client.Client, secret *corev1.Secret, owner metav1.OwnerReference) error {
```

Comment:
this doesn't necessarily seem like the best place for this function to live, especially given that we're having to add all those k8s client imports in here. Would it make sense to add it to one of the controllers? Or even make it its own controller, just like the azurejson controllers? Is this something we expect to be shared between multiple clusters or is there one per AzureCluster?

Reply:
I can try to push the ASO secrets reconciliation in a controller. There are few aspects to keep in mind here when creating/reconciling an ASO secret.
If a a new AzureCluster or ManagedControlPlane object is created we need to check few things:
- if it has an AzureClusterIdentity configured, create an ASO secret having this AzureClusterIdentity as the owner.
- if not, check if the subscriptionID is set. If subscriptionID not set, don't create any ASO secrets. If set and different than the one configured at the capz controller level, create an ASO secret having the cluster as the owner.

The new controller should watch for newly created  AzureCluster or ManagedControlPlane objects and react accordingly.

Reply:
>Is this something we expect to be shared between multiple clusters or is there one per AzureCluster?

Yes, there might be cases when an ASO secret might be shared between multiple clusters like:
- when they don't have any AzureClusterIdentity configured and have the same subscription ID and it is different than the once configured at the ASO controller level. 
- when they are using the same subscriptionID and reference the same AzureClusterIdentity.

Reply:
> If a a new AzureCluster or ManagedControlPlane object is created we need to check few things:

Since these are things the current approach already does, I don't see a reason why doing the same things from a separate controller would necessarily be much different aside from the better architectural consistency that Cecile points out.

But if the ASO secrets are managed from a separate controller, is there a mechanism we can utilize to block the main controller from its normal flow until the ASO secret controller has done its job for a particular cluster? If not, I could see spurious errors coming up if there's a race between the secret controller creating the secret and the normal controller creating a resource referencing that secret.

One option I'm not sure has been said explicitly out loud yet is to never rely on the default ASO credentials and instead always create an ASO secret for each cluster. That would obviously mean creating more secrets than strictly necessary, but I wonder if that tradeoff is worth simplifying the logic from

> * if it has an AzureClusterIdentity configured, create an ASO secret having this AzureClusterIdentity as the owner.
> * if not, check if the subscriptionID is set. If subscriptionID not set, don't create any ASO secrets. If set and different than the one configured at the capz controller level, create an ASO secret having the cluster as the owner.

to something closer to

> * initialize a `ClusterScope` (which handles whether or not an AzureClusterIdentity exists)
> * create an ASO secret having this AzureCluster as the owner from the credentials stored in the `ClusterScope`.

Reply:
> But if the ASO secrets are managed from a separate controller, is there a mechanism we can utilize to block the main controller from its normal flow until the ASO secret controller has done its job for a particular cluster

Yes, we could requeue the controller until the secret (or whatever outcome we're waiting for) is ready, something like:  https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremachine_controller.go#L265. We'd also want to be watching the secrets so that the controller gets requeued automatically when a secret is created (and we could use a watch filter/predicate to only enqueue when it's specifically an ASO secret for that cluster).

That being said I like @nojnhuh's suggestion to simplify the logic. It does seem like the special case handling of "subscription ID is set and is different than the controller one" is a bit of an edge case and we're duplicating logic that's already in `ClusterScope` that handles getting the right credentials based on subscriptionID and identity ref

Reply:
Thanks for the feedback. Based on the discussion from this thread, I went ahead and did an overhaul by simplifying and moving the logic inside a dedicated capz controller.